src/dynamics/mpas/driver/dyn_mpas_subdriver.F90

```diff
@@ -1573,6 +2370,611 @@ subroutine dyn_mpas_exchange_halo(self, field_name)
+        call self % get_variable_pointer(config_dt, 'cfg', 'config_dt')
+
+        call atm_mpas_init_block(self % domain_ptr % dminfo, self % domain_ptr % streammanager, self % domain_ptr % blocklist, &
+            mpas_pool, config_dt)
+
+        nullify(mpas_pool)
+        nullify(config_dt)
+
+        call self % get_variable_pointer(xtime_pointer, 'state', 'xtime', time_level=1)
+
+        xtime_pointer = date_time
+
+        nullify(xtime_pointer)
+
+        ! Initialize `initial_time` in the second time level. We need to do this manually because initial states
+        ! are read into time level 1, and if we write anything from time level 2, `initial_time` will be invalid.
+        call self % get_variable_pointer(initial_time_1_pointer, 'state', 'initial_time', time_level=1)
+        call self % get_variable_pointer(initial_time_2_pointer, 'state', 'initial_time', time_level=2)
+
+        initial_time_2_pointer = initial_time_1_pointer
+
+        ! Set time units to CF-compliant "seconds since <date and time>".
+        call self % get_pool_pointer(mpas_pool, 'state')
+
+        call mpas_pool_get_field(mpas_pool, 'Time', field_0d_real, timelevel=1)
+
+        if (.not. associated(field_0d_real)) then
+            call self % model_error('Failed to find variable "Time"', subname, __LINE__)
+        end if
+
+        call mpas_modify_att(field_0d_real % attlists(1) % attlist, 'units', &
+            'seconds since ' // mpas_string_replace(initial_time_1_pointer, '_', ' '), ierr=ierr)
+
+        if (ierr /= 0) then
+            call self % model_error('Failed to set time units', subname, __LINE__)
+        end if
+
+        nullify(initial_time_1_pointer)
+        nullify(initial_time_2_pointer)
+        nullify(mpas_pool)
+        nullify(field_0d_real)
+
+        call exchange_halo_group(self % domain_ptr, 'initialization:pv_edge,ru,rw', ierr=ierr)
+
+        if (ierr /= 0) then
+            call self % model_error('Failed to exchange halo layers for group "initialization:pv_edge,ru,rw"', subname, __LINE__)
+        end if
+
+        call self % debug_print('Initializing dynamics')
+
+        ! Prepare dynamics for time integration.
+        call mpas_atm_dynamics_init(self % domain_ptr)
+
+        call self % debug_print(subname // ' completed')
+
+        call self % debug_print('Successful initialization of MPAS dynamical core')
+    contains
+        !> Test if `a` is divisible by `b`, where `a` and `b` are both reals.
+        !> (KCW, 2024-05-25)
+        pure function almost_divisible(a, b)
+            real(rkind), intent(in) :: a, b
+            logical :: almost_divisible
+
+            real(rkind) :: error_tolerance
+
+            error_tolerance = epsilon(1.0_rkind) * max(abs(a), abs(b))
+
+            if (almost_equal(mod(abs(a), abs(b)), 0.0_rkind, absolute_tolerance=error_tolerance) .or. &
+                almost_equal(mod(abs(a), abs(b)), abs(b), absolute_tolerance=error_tolerance)) then
```

Comment:
Could you clarify why this second `almost_equal` is needed?

Reply:
Ok, stepping away from code for a bit can do wonders. If `a` is slightly less an an integer multiple of `b` we need this check.